### PR TITLE
Hide top widget and tabbar in fullscreen

### LINKED ISSUE #571: Fix fullscreen if there is no current webview (zim)
The fullscreen functionality works by putting the current (zim) webview fullscreen.
But there is no safeguard if there is no current zim displayed and we end by passing a nullptr to the FullScreen class.

We have a technical bug (using a nullptr) and a functional point to fix (what to do if there is no current webview)

The simpler seems to deactivate the fullscreen if there is no current zim.
But then we have to deactivate the action, not just add a safeguard.
And we probably need to test corner use case (closing the webview while we are fullscreen, ...)

Or we may want to keep the fullscreen functionality if there is no current zim, but we must decide what is the correct behavior.

### PR DESCRIPTION
**Pull request by juuz0**
This change hides the topwidget and tabbar when fullscreen shortcut/button is pressed.
If the mouse is taken to the top of screen, they are shown till the time it is there.
Fixes #571

### CODE REVIEW

**Review by kelson42 — APPROVED** [review]
From a user perspective, the PR fix the problem the right way.

**Comment by kelson42** [comment]
@veloman-yunkan Considering that this PR is already more than a week old that @mgautierfr will be away for a while, could you please review the code of this PR?

File: src/mainwindow.cpp [hunk 1]
```diff
@@ -67,6 +68,8 @@ MainWindow::MainWindow(QWidget *parent) :
 
     mp_ui->contentmanagerside->setContentManager(app->getContentManager());
     mp_ui->sideBar->setCurrentWidget(mp_ui->contentmanagerside);
+    QApplication::instance()->installEventFilter(this);
```
[the remark this answers is not in the document]
Comment (juuz0) [reply src/mainwindow.cpp:71]: I tried mouse move events directly earlier, but they just never fired without a mouse click (mouse tracking was enabled).
[Here's a similar problem](https://stackoverflow.com/questions/32714105/mousemoveevent-is-not-called).

[the remark this answers is not in the document]
Comment (veloman-yunkan) [reply src/mainwindow.cpp:71]: Indeed. It looks that the mouse events are delivered to the innermost widget, necessitating the event filtering approach. However, I still recommend having event filtering enabled only while in full-screen mode).

Besides, there is no sense in calling `setMouseTracking()` on the next line.

File: src/mainwindow.cpp [hunk 2]
```diff
@@ -75,12 +76,44 @@ MainWindow::~MainWindow()
...
+}
+
+bool MainWindow::eventFilter(QObject* /*object*/, QEvent* event)
+{
+    if (event->type() == QEvent::MouseMove && isFullScreen())
+    {
+        const auto mouseEvent = static_cast<QMouseEvent*>(event);
+        if (mouseEvent->y() == 0) {
+            showTabAndTop();
+        } else if(mouseEvent->y() >= 150) {
```
Comment (veloman-yunkan) [inline src/mainwindow.cpp:108]: How did you select this value? In my environment it is noticeably below the bottom of the tab bar.

Comment (juuz0) [reply src/mainwindow.cpp:108]: The noticeably below was intentional - that's how Firefox does it also. Though value selection was not optimal, found another way this time.

File: src/mainwindow.cpp [hunk 3]
```diff
@@ -75,12 +76,45 @@ MainWindow::~MainWindow()
...
+}
+
+void MainWindow::showTabAndTop() {
+    getTabBar()->show();
+    getTopWidget()->show();
+}
+
+bool MainWindow::eventFilter(QObject* /*object*/, QEvent* event)
+{
+    const int tabRegion = getTabBar()->height() + getTopWidget()->height() + 30;
```
Comment (veloman-yunkan) [inline src/mainwindow.cpp:103]: Compute this value inside the body of the `if` statement.

File: src/mainwindow.cpp [hunk 4]
```diff
@@ -75,12 +76,45 @@ MainWindow::~MainWindow()
...
+    if (event->type() == QEvent::MouseMove && isFullScreen())
+    {
+        const auto mouseEvent = static_cast<QMouseEvent*>(event);
+        if (mouseEvent->y() == 0) {
+            showTabAndTop();
+        } else if(mouseEvent->y() >= tabRegion) {
+            hideTabAndTop();
+        }
```
Comment (veloman-yunkan) [inline src/mainwindow.cpp:111]: Please state in a comment the (so far) implicit assumption that calling `hide()` on a hidden widget or `show()` on a non-hidden widget is a cheap no-op.

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]

File: src/mainwindow.cpp [hunk 5]
```diff
@@ -75,12 +76,46 @@ MainWindow::~MainWindow()
...
+    getTopWidget()->show();
+}
+
+bool MainWindow::eventFilter(QObject* /*object*/, QEvent* event)
+{
+    if (event->type() == QEvent::MouseMove && isFullScreen())
+    {
+        const auto mouseEvent = static_cast<QMouseEvent*>(event);
+        const int tabRegion = getTabBar()->height() + getTopWidget()->height() + 30;
+        // We don't have to check for visibilty as calling hide() on a hidden widget, or show() on a non-hidden widget is a no-op
```
Comment (veloman-yunkan) [inline src/mainwindow.cpp:107]: Since you left out the word "cheap" I would like to clarify that keeping `show()`ing a visible widget on every mouse move in the bigger part of the screen is acceptable only if that operation has negligible performance overhead rather than just being [idempotent](https://en.wikipedia.org/wiki/Idempotence). An example to the contrary would be if `QWidget::show()` unconditionally enqueued a repaint event.

**Review by veloman-yunkan — APPROVED** [review]

File: src/mainwindow.cpp (hunk 5 above)
Comment (juuz0) [reply src/mainwindow.cpp:107]: > if that operation has negligible performance overhead rather than just being [idempotent](https://en.wikipedia.org/wiki/Idempotence).

Noted, thank you.

> An example to the contrary would be if QWidget::show() unconditionally enqueued a repaint event.

True, I checked into code and found out that there's a [check in setVisible](https://code.woboq.org/qt5/qtbase/src/widgets/kernel/qwidget.cpp.html#_ZN7QWidget10setVisibleEb) already. That should be fine to qualify it as a no-op I think?

Comment (veloman-yunkan) [reply src/mainwindow.cpp:107]: Yes.
